File: apps/companion/android/src/main/java/com/eonreality/fieldiq/MetaGlassesModule.kt
// MetaGlassesModule.kt
//
// React Native bridge wrapping the Meta Wearables Device Access Toolkit for
// Android. See https://github.com/facebook/meta-wearables-dat-android.
//
// The DAT SDK is added via the generated build.gradle after `expo prebuild` on
// the dev machine.  Until that lands, methods return deterministic stubs so the
// JS layer can still exercise the flow.
//
// TODO(meta-sdk-link):
//   1. Add `implementation 'com.facebook.meta.wearables:dat:<version>'` to
//      android/app/build.gradle.
//   2. Replace the stub bodies with calls into `com.facebook.meta.wearables.dat.*`.
//   3. Emit `MetaGlasses.connection` events through DeviceEventManagerModule.

package com.eonreality.fieldiq

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.modules.core.DeviceEventManagerModule
// import com.facebook.meta.wearables.dat.DeviceManager  // uncomment after gradle dep is added.

class MetaGlassesModule(private val ctx: ReactApplicationContext) :
  ReactContextBaseJavaModule(ctx) {

  override fun getName(): String = "MetaGlasses"

  private fun emit(status: String) {
    ctx
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit("MetaGlasses.connection", status)
  }

  @ReactMethod
  fun pairDevice(promise: Promise) {
    // TODO(meta-sdk-link): DeviceManager.pair(ctx) { result -> ... }
    emit("connecting")
    val device = Arguments.createMap().apply {
      putString("id", java.util.UUID.randomUUID().toString())
      putString("serial", "RAYBAN-DISPLAY-DEV")
      putString("model", "meta-ray-ban-display")
      putInt("batteryPercent", 92)
    }
    emit("paired")
    promise.resolve(device)
  }

  @ReactMethod
  fun getPairedDevice(promise: Promise) {
    // TODO(meta-sdk-link): DeviceManager.getPaired(ctx) { device -> ... }
    promise.resolve(null)
  }

  @ReactMethod
  fun capturePhoto(promise: Promise) {
    // TODO(meta-sdk-link): glassesCamera.capturePhoto { jpeg ->
    //   val map = Arguments.createMap()
    //   map.putString("base64", Base64.encodeToString(jpeg, Base64.NO_WRAP))
    //   map.putInt("width", w); map.putInt("height", h)
    //   map.putString("capturedAt", java.time.Instant.now().toString())
    //   promise.resolve(map)
    // }
    promise.reject(
      "not_linked",
      "Meta Wearables DAT SDK is not linked yet. Run `expo prebuild` and add the gradle dep.",
    )
  }

  @ReactMethod
  fun getBatteryLevel(promise: Promise) {
    promise.resolve(0)
  }

  // For RN's NativeEventEmitter contract.
  @ReactMethod fun addListener(eventName: String) { /* no-op */ }
  @ReactMethod fun removeListeners(count: Int) { /* no-op */ }
}

File: apps/companion/ios/MetaGlassesModule.swift
// MetaGlassesModule.swift
//
// React Native bridge wrapping the Meta Wearables Device Access Toolkit for iOS.
// See https://github.com/facebook/meta-wearables-dat-ios.
//
// The DAT SDK is added via CocoaPods (`pod 'MetaWearablesDAT'`) after `expo
// prebuild` on a Mac. Until then the methods short-circuit to deterministic
// responses so the JS layer can still exercise the flow.
//
// TODO(meta-sdk-link):
//   1. Add `pod 'MetaWearablesDAT'` to the generated Podfile or vendor the
//      framework into ios/Frameworks/MetaWearablesDAT.xcframework.
//   2. Replace the stub bodies below with calls into `MetaWearables.DeviceManager`
//      and `MetaWearables.Camera` (see https://wearables.developer.meta.com/docs).
//   3. Forward connection-status changes through the JS event emitter
//      (`MetaGlasses.connection`).

import Foundation
import React
// import MetaWearablesDAT  // uncomment after `pod install`

@objc(MetaGlasses)
class MetaGlasses: RCTEventEmitter {

  private var listenerCount = 0
  // private var deviceManager: MWDeviceManager?  // uncomment once SDK is linked.

  override static func requiresMainQueueSetup() -> Bool { false }
  override func supportedEvents() -> [String]! { ["MetaGlasses.connection"] }

  override func startObserving() { listenerCount += 1 }
  override func stopObserving()  { listenerCount = max(0, listenerCount - 1) }

  @objc func pairDevice(_ resolve: @escaping RCTPromiseResolveBlock,
                        reject: @escaping RCTPromiseRejectBlock) {
    // TODO(meta-sdk-link): deviceManager?.pair { result in ... }
    sendEvent(withName: "MetaGlasses.connection", body: "connecting")
    DispatchQueue.main.asyncAfter(deadline: .now() + 0.4) {
      self.sendEvent(withName: "MetaGlasses.connection", body: "paired")
      resolve([
        "id": UUID().uuidString,
        "serial": "RAYBAN-DISPLAY-DEV",
        "model": "meta-ray-ban-display",
        "batteryPercent": 92
      ])
    }
  }

  @objc func getPairedDevice(_ resolve: @escaping RCTPromiseResolveBlock,
                             reject: @escaping RCTPromiseRejectBlock) {
    // TODO(meta-sdk-link): deviceManager?.getPaired { ... }
    resolve(NSNull())
  }

  @objc func capturePhoto(_ resolve: @escaping RCTPromiseResolveBlock,
                          reject: @escaping RCTPromiseRejectBlock) {
    // TODO(meta-sdk-link): camera.capturePhoto { result in
    //   resolve(["base64": result.base64, "width": result.width, "height": result.height,
    //            "capturedAt": ISO8601DateFormatter().string(from: Date())])
    // }
    reject("not_linked",
           "Meta Wearables DAT SDK is not linked yet. Run `expo prebuild` on a Mac and add the pod.",
           nil)
  }

  @objc func getBatteryLevel(_ resolve: @escaping RCTPromiseResolveBlock,
                             reject: @escaping RCTPromiseRejectBlock) {
    resolve(0)
  }
}
